packages/loopring_v3/contracts/amm/libamm/AmmTransactionReceiver.sol

```diff
@@ -94,19 +105,59 @@ library AmmTransactionReceiver
             signature.offset := add(signature.offset, 0x20)
         }
         if (txType == AmmData.PoolTxType.JOIN) {
+            S.approveAmmUpdates(ctx, true);
             S.processJoin(
                 ctx,
                 abi.decode(data, (AmmData.PoolJoin)),
                 signature
             );
+            S.approveAmmUpdates(ctx, false);
         } else if (txType == AmmData.PoolTxType.EXIT) {
+            S.approveAmmUpdates(ctx, true);
             S.processExit(
                 ctx,
                 abi.decode(data, (AmmData.PoolExit)),
                 signature
             );
+            S.approveAmmUpdates(ctx, false);
+        } else if (txType == AmmData.PoolTxType.SET_VIRTUAL_BALANCES) {
+            S.approveAmmUpdates(ctx, true);
+            processSetVirtualBalances(
+                ctx,
+                abi.decode(data, (AmmData.PoolVirtualBalances))
+            );
+            S.approveAmmUpdates(ctx, false);
+        } else if (txType == AmmData.PoolTxType.DEPOSIT) {
+            S.processDeposit(
+                ctx,
+                abi.decode(data, (AmmData.PoolDeposit))
+            );
+         } else if (txType == AmmData.PoolTxType.WITHDRAW) {
+             S.processWithdrawal(
+                ctx,
+                abi.decode(data, (AmmData.PoolWithdrawal))
+            );
         } else {
             revert("INVALID_POOL_TX_TYPE");
         }
     }
+
+    function processSetVirtualBalances(
```

Comment:
Can we move this method to its own library file (AmmVirtualBalanceProcess.sol)

Reply:
Sure.